src/Disks/ObjectStorages/ObjectStorageIteratorAsync.cpp

```diff
@@ -76,10 +76,12 @@ void IObjectStorageIteratorAsync::next()
 {
     std::lock_guard lock(mutex);
 
+    if (is_finished)
+        return;
+
+    ++current_batch_iterator;
```

Comment:
Here was an error. The iterator was able to iterate only over a single batch.

Reply:
@alexey-milovidov please extract bugfixes for async iterator into a separate PR so that we can backport them to 24.6, unless we can backport this PR as a whole. (Since 24.6 this iterator is used for s3/etc table engines in `StorageObjectStorage.cpp`)

Reply:
I checked, and StorageObjectStorage only used "batch" methods from the iterator (btw, I don't know why), but the bug was only in the `next`/`current` interface of the iterator. It means that it is likely not affected.